File: 144. Binary Tree Preorder Traversal/Solution.py
# LeetCode #144: Binary Tree Preorder Traversal

'''
Given a binary tree, return the preorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},
   1
    \
     2
    /
   3
return [1,2,3].

Note: Recursive solution is trivial, could you do it iteratively?
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution: # Recursion
    def preorderTraversal(self, root):
        res = []
        if root is None:
            return res
            
        res.append(root.val)
        res.extend(self.preorderTraversal(root.left))
        res.extend(self.preorderTraversal(root.right))
        
        return res

'''
class Solution: # Iteration
    # @param {TreeNode} root
    # @return {integer[]}
    def preorderTraversal(self, root):
        if root == None:
            return []
            
        stack = [root]
        res = []
        
        while len(stack) > 0:
            tmp = stack.pop()
            res.append(tmp.val)
            
            if tmp.right != None:
                stack.append(tmp.right)
            if tmp.left != None:
                stack.append(tmp.left)
        return res
'''           

File: 144. Binary Tree Preorder Traversal/Solution.java
// LeetCode #144: Binary Tree Preorder Traversal

/*
Given a binary tree, return the preorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},
   1
    \
     2
    /
   3
return [1,2,3].

Note: Recursive solution is trivial, could you do it iteratively?
*/

/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        if (root == null)
            return res;
        
        Stack<TreeNode> s = new Stack<TreeNode>();
        s.push(root);
        
        while (!s.isEmpty()) {
            TreeNode tmp = s.pop();
            res.add(tmp.val);
                
            if (tmp.right != null)
                s.push(tmp.right);
            if (tmp.left != null)
                s.push(tmp.left);
        }
        return res;
    }
}